File: youtube/src/main/java/com/giufu/youtube/MainActivity.java
package com.giufu.youtube;

import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.giufu.youtube.utils.GlassGestureDetector;
import java.util.List;


public class MainActivity extends AppCompatActivity implements GlassGestureDetector.OnGestureListener {
    private static final int REQUEST_CODE = 999;
    private GlassGestureDetector glassGestureDetector;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getSupportActionBar().hide();
        setContentView(R.layout.activity_main);
        glassGestureDetector = new GlassGestureDetector(this, this);
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            hideSystemUI();
        }
    }

    private void detectSpeech() {
        final Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
        intent.putExtra(
                RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
        startActivityForResult(intent, REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == RESULT_OK) {
            final List<String> results = data.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS);
            Log.d("app", "results: " + results.toString());
            if (results != null && results.size() > 0 && !results.get(0).isEmpty()) {
                String speechResult = results.get(0);
                Intent intent = new Intent(this,ResultsActivity.class);
                intent.putExtra("query",speechResult);
                startActivity(intent);
                Log.d("app", speechResult);
            }
        } else {
            Log.d("app", "Result not OK");
        }
    }

    private void hideSystemUI() {
        View decorView = getWindow().getDecorView();
        decorView.setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_IMMERSIVE
                        | View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_FULLSCREEN);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        return glassGestureDetector.onTouchEvent(ev) || super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onGenericMotionEvent(MotionEvent ev) {
        return glassGestureDetector.onTouchEvent(ev) || super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onGesture(GlassGestureDetector.Gesture gesture) {
        switch (gesture) {
            case TAP:
                detectSpeech();
                return true;
            default:
                return false;
        }
    }
}

File: youtube_explorer/src/main/java/com/giufu/youtube_explorer/MainActivity.kt
package com.giufu.youtube_explorer

import android.app.Activity
import android.content.Intent
import android.speech.RecognizerIntent
import android.view.*

class MainActivity : Activity() {
    private val SPEECH_REQUEST = 0
    private var stopped: Boolean = false
    private var shouldFinishOnMenuClose = true

    private fun displaySpeechRecognizer() {
        shouldFinishOnMenuClose = false;
        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH)
        startActivityForResult(intent, SPEECH_REQUEST)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        openOptionsMenu()
    }

    override fun onActivityResult(
        requestCode: Int, resultCode: Int,
        data: Intent) {
        if (requestCode == SPEECH_REQUEST && resultCode == RESULT_OK) {
            val results: List<String> = data.getStringArrayListExtra(
                RecognizerIntent.EXTRA_RESULTS)
            val spokenText = results[0]
            val intent = Intent(this, ResultsActivity::class.java)
            intent.putExtra("query", spokenText)
            startActivity(intent)
            stopped = true
        }
        shouldFinishOnMenuClose = true
        finish()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_activity, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        shouldFinishOnMenuClose = true
        when (item.itemId) {
            R.id.voice_search -> {
                stopped = true
                displaySpeechRecognizer()
                return true
            }
            R.id.scan_qr -> {
                val intent = Intent(this, QrScannerActivity::class.java)
                startActivity(intent)
                return true
            }
            R.id.action_stop -> {
                //stopService(Intent(this, LiveCardService::class.java))
                finish()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onOptionsMenuClosed(menu: Menu) {
        super.onOptionsMenuClosed(menu)
        if (shouldFinishOnMenuClose) {
            finish();
        }
    }
}